pkg/ring/ring_test.go

```diff
@@ -243,6 +243,181 @@ func TestRing_Get_ZoneAwareness(t *testing.T) {
 	}
 }
 
+func TestRing_GetAllHealthy(t *testing.T) {
+	const heartbeatTimeout = time.Minute
+	now := time.Now()
+
+	tests := map[string]struct {
+		ringInstances           map[string]IngesterDesc
+		expectedErrForRead      error
+		expectedSetForRead      []string
+		expectedErrForWrite     error
+		expectedSetForWrite     []string
+		expectedErrForReporting error
+		expectedSetForReporting []string
+	}{
+		"should return error on empty ring": {
+			ringInstances:           nil,
+			expectedErrForRead:      ErrEmptyRing,
+			expectedErrForWrite:     ErrEmptyRing,
+			expectedErrForReporting: ErrEmptyRing,
+		},
+		"should return all healthy instances for the given operation": {
+			ringInstances: map[string]IngesterDesc{
+				"instance-1": {Addr: "127.0.0.1", State: ACTIVE, Timestamp: now.Unix()},
+				"instance-2": {Addr: "127.0.0.2", State: PENDING, Timestamp: now.Add(-10 * time.Second).Unix()},
+				"instance-3": {Addr: "127.0.0.3", State: JOINING, Timestamp: now.Add(-20 * time.Second).Unix()},
+				"instance-4": {Addr: "127.0.0.4", State: LEAVING, Timestamp: now.Add(-30 * time.Second).Unix()},
+				"instance-5": {Addr: "127.0.0.5", State: ACTIVE, Timestamp: now.Add(-2 * time.Minute).Unix()},
+			},
+			expectedSetForRead:      []string{"127.0.0.1", "127.0.0.2", "127.0.0.4"},
```

Comment:
Funny that PENDING can be read from, but JOINING cannot (ingester is in PENDING state before it enters JOINING state). Anyway, that's not related to this PR.